### DISCUSSION TITLE
Make IUser work like ISite

### DISCUSSION THREAD

**Discussion by infofromca** [OrchardCMS/OrchardCore#15879] [Q&A]
I need to expand IUser like ISite in this way

```
 var container = await _siteService.GetSiteSettingsAsync();
            container.Properties[nameof(OpenIdClientSettings)] = JObject.FromObject(settings);
            await _siteService.UpdateSiteSettingsAsync(container);

```
I used this code:
` var container = await _userStore.FindByNameAsync(User.Identity.Name, default(CancellationToken));
           container.Properties[nameof(UserSearchProperty)] = JObject.FromObject(userSearchProperty);
            await _userStore.UpdateAsync(container, default(CancellationToken));
            return View(model);`
but it does not work.
I should comment out
  // container.Properties[nameof(UserSearchProperty)] = JObject.FromObject(userSearchProperty);

**Comment by Skrypt** [comment]
There's an example on how to extend the IUser if you look at how I did with User TimeZones.

**Comment by infofromca** [comment]
I think you mean 
  var user = await _userManager.FindByNameAsync(userName) as User;
but in this way, we should cast two times . (one for get, one for update).
meanwhile, I think the whole application should be consistent.
I even think all of the entities come from IEntity.

**Comment by sebastienros** [comment]
There is an example of custom user profile in the Demo module, have you looked at it?

**Comment by petedavis** [comment]
It does work like ISite, just have to use `User` rather than `IUser` as `User` extends `Entity`, but IUser doesnt extend `IEntity`

See [UserProfileDisplayDriver](https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Demo/Drivers/UserProfileDisplayDriver.cs)

@sebastienros  is `IUser` not extending the `IEntity` interface a bug? I think it would be a 10 character pull request.

**Comment by Skrypt** [comment]
I think the goal is to abstract the User data layer for security purpose. Also, if IUser was extending directly the Entity class then we could end up with multiple implementation of the User data migrations and store. So, right now, the suggested way to extend a User is by using User properties data. Does that make sense?

The IUser interface has just one field and it's the UserName which is the only thing required for a user to really work in Orchard. Also, the logic is : no need to have a User store for a User to work...

```JSON
{"Id":61,"UserName":"admin","NormalizedUserName":"ADMIN","Email":"email@email.com","NormalizedEmail":"EMAIL@EMAIL.COM","PasswordHash":"somehashhere","SecurityStamp":"somesecuritystamp","EmailConfirmed":true,"ResetToken":null,"RoleNames":{"$type":"System.String[], System.Private.CoreLib","$values":["Administrator"]},"LoginInfos":[],
"Properties":{
//HERE IS WHERE YOU EXTEND User
}}
```

**Comment by sebastienros** [comment]
The issue is not about adding 10 chars, but the impact of 10 chars. By implementing IEntity we expect any service providing an IUser implementation to depend on it and support it. It might be the right thing to do, but it needs to be thought about.
